File: lv-preisrechner/backend/app/schemas/pricing.py
"""Pydantic-Schemas für die Pricing-Foundation (B+1)."""

from __future__ import annotations

from datetime import date, datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator


class PricelistStatusSchema(str, Enum):
    PENDING_PARSE = "PENDING_PARSE"
    PARSING = "PARSING"
    PARSED = "PARSED"
    REVIEWED = "REVIEWED"
    APPROVED = "APPROVED"
    ARCHIVED = "ARCHIVED"
    ERROR = "ERROR"


# ---------------------------------------------------------------------------
# SupplierPriceList
# ---------------------------------------------------------------------------
class _DateRangeMixin(BaseModel):
    """Gueltigkeits-Zeitraum-Validierung."""

    @model_validator(mode="after")
    def _check_dates(self):
        vf = getattr(self, "valid_from", None)
        vu = getattr(self, "valid_until", None)
        if vf and vu and vu < vf:
            raise ValueError("valid_until muss >= valid_from sein")
        return self


class SupplierPriceListCreate(_DateRangeMixin):
    """Upload-Request. Datei-Metadaten werden separat als Multipart angenommen."""

    supplier_name: str = Field(..., min_length=1, max_length=200)
    supplier_location: str | None = Field(None, max_length=200)
    list_name: str = Field(..., min_length=1, max_length=200)
    valid_from: date
    valid_until: date | None = None


class SupplierPriceListOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    tenant_id: str
    supplier_name: str
    supplier_location: str | None
    list_name: str
    valid_from: date
    valid_until: date | None
    source_file_path: str
    source_file_hash: str
    status: PricelistStatusSchema
    parse_error: str | None
    entries_total: int | None
    entries_reviewed: int | None
    is_active: bool
    uploaded_by_user_id: str
    uploaded_at: datetime
    approved_by_user_id: str | None
    approved_at: datetime | None


# ---------------------------------------------------------------------------
# SupplierPriceEntry
# ---------------------------------------------------------------------------
class SupplierPriceEntryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    pricelist_id: str
    tenant_id: str
    article_number: str | None
    manufacturer: str | None
    product_name: str
    category: str | None
    subcategory: str | None
    price_net: float
    currency: str
    unit: str
    package_size: float | None
    package_unit: str | None
    pieces_per_package: int | None
    effective_unit: str
    price_per_effective_unit: float
    attributes: dict[str, Any]
    source_page: int | None
    parser_confidence: float
    needs_review: bool
    reviewed_by_user_id: str | None = None
    reviewed_at: datetime | None = None
    correction_applied: bool = False


class SupplierPriceEntryUpdate(BaseModel):
    """Partial-Update fuer einen Entry. Alle Felder optional.

    Der Server erkennt Aenderungen, setzt correction_applied=True und
    markiert den Reviewer. Wenn needs_review von True auf False wechselt,
    wird der entries_reviewed-Counter auf der Parent-Pricelist
    inkrementiert (umgekehrt dekrementiert).
    """

    model_config = ConfigDict(extra="forbid")

    article_number: str | None = Field(None, max_length=100)
    manufacturer: str | None = Field(None, max_length=200)
    product_name: str | None = Field(None, min_length=1, max_length=500)
    category: str | None = Field(None, max_length=200)
    subcategory: str | None = Field(None, max_length=200)
    price_net: float | None = Field(None, gt=0)
    unit: str | None = Field(None, min_length=1, max_length=50)
    effective_unit: str | None = Field(None, min_length=1, max_length=50)
    price_per_effective_unit: float | None = Field(None, gt=0)
    package_size: float | None = Field(None, gt=0)
    package_unit: str | None = Field(None, max_length=50)
    pieces_per_package: int | None = Field(None, gt=0)
    attributes: dict[str, Any] | None = None
    needs_review: bool | None = None


class SupplierPriceListDetail(SupplierPriceListOut):
    entries: list[SupplierPriceEntryOut] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# TenantPriceOverride
# ---------------------------------------------------------------------------
class TenantPriceOverrideCreate(_DateRangeMixin):
    article_number: str = Field(..., min_length=1, max_length=100)
    manufacturer: str | None = Field(None, max_length=200)
    override_price: float = Field(..., gt=0)
    unit: str = Field(..., min_length=1, max_length=50)
    valid_from: date
    valid_until: date | None = None
    notes: str | None = None


class TenantPriceOverrideOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    tenant_id: str
    article_number: str
    manufacturer: str | None
    override_price: float
    unit: str
    valid_from: date
    valid_until: date | None
    notes: str | None
    created_by_user_id: str
    created_at: datetime


# ---------------------------------------------------------------------------
# TenantDiscountRule
# ---------------------------------------------------------------------------
class TenantDiscountRuleCreate(_DateRangeMixin):
    supplier_name: str = Field(..., min_length=1, max_length=200)
    discount_percent: float = Field(..., ge=0, le=100)
    category: str | None = Field(None, max_length=200)
    valid_from: date
    valid_until: date | None = None
    notes: str | None = None


class TenantDiscountRuleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    tenant_id: str
    supplier_name: str
    discount_percent: float
    category: str | None
    valid_from: date
    valid_until: date | None
    notes: str | None
    created_by_user_id: str
    created_at: datetime


# ---------------------------------------------------------------------------
# ProductCorrection / Review-Workflow (B+4.4 P4)
# ---------------------------------------------------------------------------
class ProductCorrectionTypeSchema(str, Enum):
    PIECES_PER_PACKAGE = "pieces_per_package"
    UNIT_OVERRIDE = "unit_override"
    PRICE_PER_EFFECTIVE_UNIT = "price_per_effective_unit"
    CONFIRMED_AS_IS = "confirmed_as_is"


class EntryReviewItem(BaseModel):
    """Ein einzelner Review-Eintrag mit allem Kontext fuers UI."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    pricelist_id: str
    article_number: str | None
    manufacturer: str | None
    product_name: str
    price_net: float
    currency: str
    unit: str
    package_size: float | None
    pieces_per_package: int | None
    effective_unit: str
    price_per_effective_unit: float
    source_page: int | None
    source_row_raw: str | None
    review_reason: str | None
    attributes: dict[str, Any]
    parser_confidence: float


class EntryReviewGroup(BaseModel):
    """Gruppe von Review-Eintraegen mit identischem review_reason."""

    review_reason: str
    count: int
    items: list[EntryReviewItem]


class EntryReviewResponse(BaseModel):
    """Top-Level-Antwort von GET /pricing/entries/{pricelist_id}/review."""

    pricelist_id: str
    total_needs_review: int
    groups: list[EntryReviewGroup]


class EntryCorrectionRequest(BaseModel):
    """POST /pricing/entries/{entry_id}/correct.

    corrected_value ist typ-abhaengig — Validierung im Service-Layer:
    - PIECES_PER_PACKAGE: {"pieces_per_package": int > 0}
    - UNIT_OVERRIDE: {"unit": str, "package_size": float | null,
                       "effective_unit": str | null}
    - PRICE_PER_EFFECTIVE_UNIT: {"price_per_effective_unit": float > 0,
                                  "effective_unit": str | null}
    - CONFIRMED_AS_IS: {} (leerer Dict reicht; setzt nur Flags)
    """

    model_config = ConfigDict(extra="forbid")

    correction_type: ProductCorrectionTypeSchema
    corrected_value: dict[str, Any] = Field(default_factory=dict)
    persist: bool = Field(
        default=True,
        description=(
            "Wenn true, wird die Korrektur in lvp_product_corrections "
            "gespeichert und beim naechsten Upload automatisch angewendet."
        ),
    )


class EntryCorrectionResponse(BaseModel):
    """Antwort nach angewandter Korrektur."""

    entry: SupplierPriceEntryOut
    correction_persisted: bool
    correction_id: str | None = None

File: lv-preisrechner/frontend/src/lib/types/pricing.ts
/**
 * TypeScript-Typen fuer die neue Pricing-API (B+1/B+2).
 *
 * Schema-Quelle: lv-preisrechner/backend/app/schemas/pricing.py
 * Nicht zu verwechseln mit den Legacy-Typen PriceList / PriceEntry in api.ts
 * (die sprechen den alten /price-lists-Endpoint).
 */

/** Status-Enum aus app/models/pricing.py:PricelistStatus */
export type PricingStatus =
  | "PENDING_PARSE"
  | "PARSING"
  | "PARSED"
  | "PARTIAL_PARSE"
  | "REVIEWED"
  | "APPROVED"
  | "ARCHIVED"
  | "ERROR";

export const PRICING_STATUS_VALUES: PricingStatus[] = [
  "PENDING_PARSE",
  "PARSING",
  "PARSED",
  "PARTIAL_PARSE",
  "REVIEWED",
  "APPROVED",
  "ARCHIVED",
  "ERROR",
];

/** Menschlich lesbare Bezeichnung + Badge-Variant je Status. */
export const PRICING_STATUS_META: Record<
  PricingStatus,
  { label: string; badge: "default" | "info" | "warning" | "success" | "danger" }
> = {
  PENDING_PARSE: { label: "Warten auf Parse", badge: "info" },
  PARSING: { label: "Wird geparst …", badge: "info" },
  PARSED: { label: "Geparst", badge: "info" },
  PARTIAL_PARSE: { label: "Teilweise geparst", badge: "warning" },
  REVIEWED: { label: "Review abgeschlossen", badge: "info" },
  APPROVED: { label: "Freigegeben", badge: "success" },
  ARCHIVED: { label: "Archiviert", badge: "default" },
  ERROR: { label: "Fehler", badge: "danger" },
};

/** B+4.5 — strukturierter Batch-Fehler aus dem Parser. */
export type ParseErrorDetail = {
  batch_idx: number;
  page_range: string;
  attempts: number;
  error_class: string;
  error_message: string;
  raw_response_file: string | null;
};

/** SupplierPriceListOut */
export type SupplierPriceList = {
  id: string;
  tenant_id: string;
  supplier_name: string;
  supplier_location: string | null;
  list_name: string;
  valid_from: string; // ISO-Date YYYY-MM-DD
  valid_until: string | null;
  source_file_path: string;
  source_file_hash: string;
  status: PricingStatus;
  parse_error: string | null;
  parse_error_details: ParseErrorDetail[] | null;
  entries_total: number | null;
  entries_reviewed: number | null;
  is_active: boolean;
  uploaded_by_user_id: string;
  uploaded_at: string; // ISO-Datetime
  approved_by_user_id: string | null;
  approved_at: string | null;
};

/** SupplierPriceEntryOut */
export type SupplierPriceEntry = {
  id: string;
  pricelist_id: string;
  tenant_id: string;
  article_number: string | null;
  manufacturer: string | null;
  product_name: string;
  category: string | null;
  subcategory: string | null;
  price_net: number;
  currency: string;
  unit: string;
  package_size: number | null;
  package_unit: string | null;
  pieces_per_package: number | null;
  effective_unit: string;
  price_per_effective_unit: number;
  attributes: Record<string, unknown>;
  source_page: number | null;
  source_row_raw?: string | null;
  parser_confidence: number;
  needs_review: boolean;
  reviewed_by_user_id?: string | null;
  reviewed_at?: string | null;
  correction_applied?: boolean;
};

/** Partial-Update-Payload fuer PATCH /entries/{id} */
export type SupplierPriceEntryUpdate = Partial<{
  article_number: string | null;
  manufacturer: string | null;
  product_name: string;
  category: string | null;
  subcategory: string | null;
  price_net: number;
  unit: string;
  effective_unit: string;
  price_per_effective_unit: number;
  package_size: number | null;
  package_unit: string | null;
  pieces_per_package: number | null;
  attributes: Record<string, unknown>;
  needs_review: boolean;
}>;

/** SupplierPriceListDetail: Liste plus optional geladene Entries */
export type SupplierPriceListDetail = SupplierPriceList & {
  entries: SupplierPriceEntry[];
};

/** Filter-Params fuer listPricelists. */
export type ListPricelistsQuery = {
  status?: PricingStatus | null;
  supplier_name?: string | null;
  active?: boolean | null;
  offset?: number;
  limit?: number;
};


// --------------------------------------------------------------------------- //
// Review / Correction (B+4.4 P4)
// --------------------------------------------------------------------------- //

/** Normierte review_reason-Werte aus app/services/pricelist_parser.py. */
export const REVIEW_REASON_LABELS: Record<string, string> = {
  bundgroesse_fehlt: "Bundgröße fehlt",
  bundpreis_vs_einzelpreis_unklar: "Bund- oder Einzelpreis unklar",
  preis_ausserhalb_korridor: "Preis außerhalb des Korridors",
  einheit_nicht_erkannt: "Einheit nicht erkannt",
  unknown: "Ohne strukturierten Grund",
};

export type EntryReviewItem = {
  id: string;
  pricelist_id: string;
  article_number: string | null;
  manufacturer: string | null;
  product_name: string;
  price_net: number;
  currency: string;
  unit: string;
  package_size: number | null;
  pieces_per_package: number | null;
  effective_unit: string;
  price_per_effective_unit: number;
  source_page: number | null;
  source_row_raw: string | null;
  review_reason: string | null;
  attributes: Record<string, unknown>;
  parser_confidence: number;
};

export type EntryReviewGroup = {
  review_reason: string;
  count: number;
  items: EntryReviewItem[];
};

export type EntryReviewResponse = {
  pricelist_id: string;
  total_needs_review: number;
  groups: EntryReviewGroup[];
};

/** Backend: ProductCorrectionType-Enum-Werte. */
export type ProductCorrectionType =
  | "pieces_per_package"
  | "unit_override"
  | "price_per_effective_unit"
  | "confirmed_as_is";

export type CorrectEntryRequest = {
  correction_type: ProductCorrectionType;
  corrected_value: Record<string, unknown>;
  persist?: boolean;
};

export type CorrectEntryResponse = {
  entry: SupplierPriceEntry;
  correction_persisted: boolean;
  correction_id: string | null;
};
